README
Meteo

  Meteo is a product that let your Plone site shows the local weather forecast
  for Spain provided by the National Institute of Meteorology
  ("www.inm.es":http://www.inm.es)

Features

 * Meteo contains one portlet with a configurable look:

   * Multiple days layout: shows the weather forecast (sky status, maximum and
     minimum temperature) for today and next days.
     
   * Single day layout: shows the weather forecast for today only (sky status
     on morning and afternoon, maximum and minimum temperature).
 
 * Meteo provide a detailed weather forecast page for the next seven days: sky
   status, rainfall probability, maximum and minimum temperature, wind status
   and maximum ultraviolet index.

 * Meteo use a cache system: the INM website will not be asked on every request
   on your site. Every two hours, the INM website will be requested for fresh
   information.
   
 * Meteo contains one configlet that will easily let you enters all the
   configuration information (location code and portlet's look). Moreover is
   possible flush cache data.
 
 * Meteo is translated in english and spanish.

CHANGELOG
Changelog for Meteo

Meteo 0.2.1 - 2008/03/26

  - Fixed a bug in Plone 3.x
  - Fixed a bug when upgrading from previous versions

Meteo 0.2.0 - 2008/03/24

  - Compatible with the updated web of Spanish State Meteorological Agency
    (Agencia Estatal de Meteorología, www.aemet.es)
  - Featuring new icons
  - New detailed prediction view
  - Updated bundled BeautifulSoup library to version 3.0.5
  - Fixed minor bugs
  
Meteo 0.1.3 - 2007/12/13

  - Added Basque translations, thanks to Mikel Larreategi
  - Replaced PNG configlet icon with GIF (IE doesn't display transparent PNG)
  - Enhanced parsing of complex names of localities
  - Fixed minor bugs

Meteo 0.1.2 - 2007/11/26

  - Added metadata files for icons
  - Added documentation for Plone 3.0 support
  - Switch from info to debug for log messages

Meteo 0.1.1 - 2007/11/05

  - Added documentation for proxy support
  - Default timeout value to connect INM web server increased to 5s

Meteo 0.1.0 - 2007/09/24

  - Initial release


config.py
# -*- coding: utf-8 -*-
#
# Meteo
# Copyright (C) 2008 GMV SGI Team <http://www.gmv-sgi.es>
#
# This program is free software; you can redistribute it and/or
# modify it under the terms of version 2 of the GNU General Public
# License as published by the Free Software Foundation.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307
# USA
#
# $Id$
#
import logging 

from Products.Archetypes.utils import DisplayList

# The product name is specified. 
PROJECTNAME = "Meteo"

# Logger
LOGGER = logging.getLogger(PROJECTNAME)

# The directory that stores skin of this product is specified. 
SKINS_DIR = 'skins'

# A global name space when this product is loaded into Python is 
# preserved. It passes to the package_home() etc.
GLOBALS = globals()

# Porlets internal names, please don't touch.
PORTLET_SINGLE   = 'single'
PORTLET_MULTIPLE = 'multiple'

# XXX: no support yet
AEMET_LANGUAGES = DisplayList((
    ('es', 'Español',    'label_lang_es'),
    ('ca', 'Catalán',    'label_lang_ca'),
    ('gl', 'Gallego',    'label_lang_gl'),
    ('va', 'Valenciano', 'label_lang_va'),
    ('eu', 'Euskera',    'label_lang_eu'),
    ('en', 'Inglés',     'label_lang_en'),
    ('fr', 'Francés',    'label_lang_fr'),
))

# Cache duration in hours
CACHE_DURATION_IN_HOURS = 2

# Timeout on connect to www.aemet.es (set 0 for no timeout, not recommended)
TIMEOUT_IN_SECONDS = 5


__init__.py
# -*- coding: utf-8 -*-
#
# Meteo
# Copyright (C) 2008 GMV SGI Team <http://www.gmv-sgi.es>
#
# This program is free software; you can redistribute it and/or
# modify it under the terms of version 2 of the GNU General Public
# License as published by the Free Software Foundation.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307
# USA
#
# $Id$
#
from AccessControl import allow_module

from Products.CMFCore import utils
from Products.CMFCore.DirectoryView import registerDirectory

from Products.Meteo import MeteoTool
from Products.Meteo.config import *

allow_module('Products.Meteo.config')

# Register our directory in Zope
registerDirectory(SKINS_DIR, GLOBALS)

def initialize(context) :
    # Initialize the tool
    utils.ToolInit(config.PROJECTNAME + ' Tool',
                   tools=(MeteoTool.MeteoTool,),
                   product_name=config.PROJECTNAME,
                   icon='tool.gif',
    ).initialize(context)


MeteoTool.py
# -*- coding: utf-8 -*-
#
# Meteo
# Copyright (C) 2008 GMV SGI Team <http://www.gmv-sgi.es>
#
# This program is free software; you can redistribute it and/or
# modify it under the terms of version 2 of the GNU General Public
# License as published by the Free Software Foundation.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307
# USA
#
# $Id$
#
import time

from AccessControl import ClassSecurityInfo
from AccessControl import getSecurityManager

from Globals import InitializeClass

from OFS.SimpleItem import SimpleItem

from Products.CMFCore import permissions
from Products.CMFCore.utils import getToolByName 
from Products.CMFCore.utils import UniqueObject

from Products.Meteo.config import *

import Meteo

import LocationsTable

class MeteoTool(UniqueObject, SimpleItem):
    """
    """
    id         = 'meteo_tool'
    title      = 'Meteo Tool'
    meta_type  = 'Meteo Tool'
    plone_tool = 1
    
    security = ClassSecurityInfo()

    security.declarePrivate("manage_afterAdd")
    def manage_afterAdd(self, item, container) :
        """
        """
        SimpleItem.manage_afterAdd(self, item, container)

        self.locationCode = ""
        self.cacheDuration = 60 * 60 * CACHE_DURATION_IN_HOURS
        self.numDaysInPortlet = 3
        self.portletType = PORTLET_MULTIPLE
        self.languageCode = "es"
        
        self.cache = {
            "date" : 0,
            "data" : {"error" : "No data"},
        }
        
    security.declarePublic("getLocationCode")
    def getLocationCode(self):
        """ """
        return self.locationCode

    security.declarePublic("getNumDaysInPortlet")
    def getNumDaysInPortlet(self):
        """ """
        return self.numDaysInPortlet
    
    security.declarePublic("getLocationUrl")
    def getLocationUrl(self):
        """ """
        return Meteo.SERVER_URL % (self.languageCode +'/-m', self.locationCode)

    security.declarePublic("getPortletType")
    def getPortletType(self):
        """ """
        return self.portletType

    security.declarePublic("getNumDaysInPortlet")
    def getMaxIndexForPortlet(self):
        """ """
        data = self.getWeatherData()
        return min(self.numDaysInPortlet, len(data["forecast"]))

    security.declareProtected(permissions.ManagePortal, "searchLocation")
    def searchLocation(self, search):
        """ """
        # XXX: ugly
        results = []
        search = search.lower()
        search = search.replace('Á','a')
        search = search.replace('á','a')
        search = search.replace('É','e')
        search = search.replace('é','e')
        search = search.replace('Í','i')
        search = search.replace('í','i')
        search = search.replace('Ó','o')
        search = search.replace('ó','o')
        search = search.replace('Ú','u')
        search = search.replace('ú','u')
        
        for location in LocationsTable.locations:
            name = location[1].lower()
            if name.find(search) != -1:
                results.append(location)
        
        return results

    security.declareProtected(permissions.ManagePortal, "searchLocationByCode")
    def searchLocationByCode(self, code):
        """ """
        for location in LocationsTable.locations:
            if location[0] == code:
                return location[1]
        
        return ""

    security.declareProtected(permissions.ManagePortal, "manageFormResults")
    def manageFormResults(self, **params) :
        """ """
        purgeCache = False

        if self.locationCode != params["locationCode"]:
            self.locationCode = params["locationCode"]
            purgeCache = True

        if self.numDaysInPortlet != params["numDaysInPortlet"]:
            self.numDaysInPortlet = params["numDaysInPortlet"]

        if self.portletType != params["portletType"]:
            self.portletType = params["portletType"]
            
        if purgeCache:
            self.flushCache()
        
        return "manageFormResults success"
    
    security.declareProtected(permissions.ManagePortal, "flushCache")
    def flushCache(self):
        """ """
        self.cache["date"] = 0
        self.cache = self.cache
        
        return "flushCache success"

    security.declareProtected(permissions.ManagePortal, "renewCache")
    def renewCache(self, timeout=0):
        """ """
        LOGGER.info("renewCache: manual update cache (timeout: %s)", timeout)
        
        result = "renewCache success"
        try:
            data = Meteo.local_weather(self.locationCode,
                                       language=self.languageCode,
                                       timeout=timeout)
            self.cache["data"] = data
            self.cache["date"] = time.time()
            self.cache = self.cache
        except:
            result = "renewCache failed"
        
        return result

    security.declarePublic("getDayOfWeek")
    def getDayOfWeek(self, date):
        """ """
        dayOfWeek = ""
        
        if date.startswith(u'lun'):
            dayOfWeek = u"Monday"
        elif date.startswith(u'mar'):
            dayOfWeek = u"Tuesday"
        elif date.startswith(u'mié'):
            dayOfWeek = u"Wednesday"
        elif date.startswith(u'jue'):
            dayOfWeek = u"Thursday"
        elif date.startswith(u'vie'):
            dayOfWeek = u"Friday"
        elif date.startswith(u'sáb'):
            dayOfWeek = u"Saturday"
        elif date.startswith(u'dom'):
            dayOfWeek = u"Sunday"

        return dayOfWeek

    security.declarePublic("getSimpleWeatherData")
    def getSimpleWeatherData(self):
        """Return a dictionnary with the following keys:
            location : the location returned by the server, or an error message.
            forecast : a list of forecast which contains for each element
                iconUrl : the url of the icon to display
                iconAlternativeText : the text of the icon
                maxTemperature : the max temperature pretty formated
                minTemperature : the min temperature pretty formated
                dayOfWeek : a little label showing the day of week
            today : a dictonary contains the following elements
                iconUrlAM
                iconUrlPM
                iconAlternativeTextAM
                iconAlternativeTextPM
                maxTemperature
                minTemperature
                dayOfWeek
        """
        data = self.getWeatherData()

        if "error" in data.keys():
            return data

        weather = {
            "location" : data["location"][0],
            "today"    : {},
            "forecast" : [],
        }
        
        portal = getToolByName(self, 'portal_url').getPortalObject()
        portal_url = portal.absolute_url()

        today = data["forecast"][0]
        weather["today"] = {
            "iconUrlAM" : "%s/meteo_icons/%s" % (portal_url,
                                                 today["estado_img"][0]),
                                                 
            "iconUrlPM" : "%s/meteo_icons/%s" % (portal_url,
                                                 today["estado_img"][1]),
                                                 
            "iconAlternativeTextAM" : today["estado_alt"][0],
            "iconAlternativeTextPM" : today["estado_alt"][1],
            "maxTemperature" : u"%s&nbsp;°C" % today["temp_max"],
            "minTemperature" : u"%s&nbsp;°C" % today["temp_min"],
            "dayOfWeek": self.getDayOfWeek(today["fecha"])
        }
        
        nowHour = int(time.strftime("%H"))

        for i in range(self.getMaxIndexForPortlet()):
            x = 0 # Por defecto se muestra predicción para antes de medio día

            # Para el día actual, mostramos la predicción de la mañana o de la
            # tarde según a que hora se realiza la petición.
            if i == 0 and nowHour >= 12:
                x = 1
            
            forecast = data["forecast"][i]
            dData = {
                "iconUrl" : "%s/meteo_icons/%s" % (portal_url,
                                                   forecast["estado_img"][x]),
                                                   
                "iconAlternativeText" : forecast["estado_alt"][x],
                "maxTemperature" : u"%s&nbsp;°C" % forecast["temp_max"],
                "minTemperature" : u"%s&nbsp;°C" % forecast["temp_min"],
                "dayOfWeek" : self.getDayOfWeek(forecast["fecha"])
            }
            weather["forecast"].append(dData)

        return weather

    security.declarePublic("getWeatherData")
    def getWeatherData(self):
        """Return the complete dictionnary returned by Meteo.py.
           More details can be found in the docstrings of local_weather()
        """
        now = time.time()
        cacheTime = self.cache["date"]
        
        if cacheTime + self.cacheDuration < now:
            # Caché ha expirado
            LOGGER.info("getWeatherData: cache has expired (cache date: %s)",
                        time.ctime(cacheTime))
            
            try:
                data = Meteo.local_weather(self.locationCode,
                                           language=self.languageCode,
                                           timeout=TIMEOUT_IN_SECONDS)
                self.cache["data"] = data
                self.cache["date"] = now
                self.cache = self.cache
            
            except IOError, e:
                # Si hay un error conectando al servidor se utiliza la cache
                LOGGER.warning("getWeatherData: IOError (%s)", e)
                data = self.cache["data"]
                self.cache["date"] = now
                self.cache = self.cache
                
            except RuntimeError, e:
                return {"error" : e}
        else:
            data = self.cache["data"]
            
        return data

    security.declareProtected(permissions.ManagePortal, "migrate")
    def migrate(self):
        """
            Check that all properties are present and correctly initialized.
            If not, install them with a safe default value.
            It is safe to call it multiple times.
        """
        checkFor = [
            ("languageCode", "es"),
            ("numDaysInPortlet", 3),
            ("portletType", PORTLET_MULTIPLE)
        ]
        rValue = ""
        for attr, default in checkFor:
            if not hasattr(self, attr):
                setattr(self, attr, default)
                rValue += "Added: %s\n" % attr
        if rValue == "":
            return "Tool was already up to date."
        return rValue + "\nTool succesfully updated!"

    security.declarePublic("adminLink")
    def adminLink(self):
        """True if admin link have to be displayed
        """
        return getSecurityManager().checkPermission('Manager', self)

InitializeClass(MeteoTool)


Meteo.py
#!/usr/bin/python
# -*- coding: utf-8 -*-
#
# Meteo
# Copyright (C) 2008 GMV SGI Team <http://www.gmv-sgi.es>
#
# This program is free software; you can redistribute it and/or
# modify it under the terms of version 2 of the GNU General Public
# License as published by the Free Software Foundation.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307
# USA
#
# $Id$
#
"""
"""

__author__ = "Sergio Martín Morillas (smartin@gmv.com)"
__version__ = "0.2.0"
__copyright__ = "Copyright (C) 2008 GMV SGI Team <http://www.gmv-sgi.es>"
__license__ = "GPLv2"

import re
import socket
import urllib2

from BeautifulSoup import BeautifulSoup, Tag

SERVER_URL = "http://www.aemet.es/%s/eltiempo/prediccion/localidades?l=%s"

########################################################
# Funciones auxiliares
########################################################

"""
<table summary="..." class="tabla_datos" cellspacing="2">
    <thead>
        <tr class="cabecera_niv1">
            <th rowspan="2"><div class="cabecera_celda">Fecha</div></th>
            <th colspan="2">mié 12</th>
            <th colspan="2">jue 13</th>
            <th colspan="2">vie 14</th>
            
            <th rowspan="2">sáb 15</th>
            <th rowspan="2">dom 16</th>
            <th rowspan="2">lun 17</th>
            <th rowspan="2">mar 18</th>
        </tr>

        <tr class="cabecera_niv2">
            <th colspan="2">pm</th>
            <th>am</th>
            <th>pm</th>
            <th>am</th>
            <th>pm</th>
        </tr>
    </thead>

    <tbody>
        <tr>
            <th class="borde_rlb_th">Estado del cielo</th>
            <td class="borde_rb" colspan="2"><img src="/imagenes/gif/estado_cielo/15.gif" title="Muy nuboso" alt="Muy nuboso"></td>
            <td class="borde_b"><img src="/imagenes/gif/estado_cielo/15.gif" title="Muy nuboso" alt="Muy nuboso"></td>
            <td class="borde_rb"><img src="/imagenes/gif/estado_cielo/15.gif" title="Muy nuboso" alt="Muy nuboso"></td>
            <td class="borde_b"><img src="/imagenes/gif/estado_cielo/15.gif" title="Muy nuboso" alt="Muy nuboso"></td>
            <td class="borde_rb"><img src="/imagenes/gif/estado_cielo/15.gif" title="Muy nuboso" alt="Muy nuboso"></td>
            <td class="borde_rb"><img src="/imagenes/gif/estado_cielo/13.gif" title="Intervalos nubosos" alt="Intervalos nubosos"></td>
            <td class="borde_rb"><img src="/imagenes/gif/estado_cielo/14.gif" title="Nuboso" alt="Nuboso"></td>
            <td class="borde_rb"><img src="/imagenes/gif/estado_cielo/14.gif" title="Nuboso" alt="Nuboso"></td>
            <td class="borde_rb"><img src="/imagenes/gif/estado_cielo/13.gif" title="Intervalos nubosos" alt="Intervalos nubosos"></td>
        </tr>

        <tr>
            <th class="borde_rlb_th">Prob. precip.(%)</th>
            <td class="borde_rb" colspan="2">0&nbsp;</td>
            <td class="borde_rb" colspan="2">0&nbsp;</td>
            <td class="borde_rb" colspan="2">0&nbsp;</td>
            <td class="borde_rb">15&nbsp;</td>
            <td class="borde_rb">5&nbsp;</td>
            <td class="borde_rb">15&nbsp;</td>
            <td class="borde_rb">25&nbsp;</td>
        </tr>

        <tr>
            <th class="borde_rlb_th">T. Máxima (ºC)</th>
            <td colspan="2" class="borde_rb"><font class="texto_rojo">22&nbsp;</font></td>
            <td colspan="2" class="borde_rb"><font class="texto_rojo">23&nbsp;</font></td>
            <td colspan="2" class="borde_rb"><font class="texto_rojo">23&nbsp;</font></td>
            <td class="borde_rb"><font class="texto_rojo">21&nbsp;</font></td>
            <td class="borde_rb"><font class="texto_rojo">20&nbsp;</font></td>
            <td class="borde_rb"><font class="texto_rojo">19&nbsp;</font></td>
            <td class="borde_rb"><font class="texto_rojo">18&nbsp;</font></td>
        </tr>

        <tr>
            <th class="borde_rlb_th">T. Mínima (ºC)</th>
            <td colspan="2" class="borde_rb"><font class="texto_azul">10&nbsp;</font></td>
            <td colspan="2" class="borde_rb"><font class="texto_azul">11&nbsp;</font></td>
            <td colspan="2" class="borde_rb"><font class="texto_azul">12&nbsp;</font></td>
            <td class="borde_rb"><font class="texto_azul">11&nbsp;</font></td>
            <td class="borde_rb"><font class="texto_azul">9&nbsp;</font></td>
            <td class="borde_rb"><font class="texto_azul">9&nbsp;</font></td>
            <td class="borde_rb"><font class="texto_azul">8&nbsp;</font></td>
        </tr>

        <tr>
            <th class="borde_rlb_th">Viento</th>
            <td colspan="2" class="borde_rb"><img src="/imagenes/gif/iconos_viento/N.gif" title="Norte" alt="Norte"></td>
            <td class="borde_b"><img src="/imagenes/gif/iconos_viento/N.gif" title="Norte" alt="Norte"></td>
            <td class="borde_rb"><img src="/imagenes/gif/iconos_viento/O.gif" title="Oeste" alt="Oeste"></td>
            <td class="borde_b"><img src="/imagenes/gif/iconos_viento/SE.gif" title="Sudeste" alt="Sudeste"></td>
            <td class="borde_rb"><img src="/imagenes/gif/iconos_viento/SE.gif" title="Sudeste" alt="Sudeste"></td>
            <td class="borde_rb"><img src="/imagenes/gif/iconos_viento/NO.gif" title="Noroeste" alt="Noroeste"></td>
            <td class="borde_rb"><img src="/imagenes/gif/iconos_viento/N.gif" title="Norte" alt="Norte"></td>
            <td class="borde_rb"><img src="/imagenes/gif/iconos_viento/NO.gif" title="Noroeste" alt="Noroeste"></td>
            <td class="borde_rb"><img src="/imagenes/gif/iconos_viento/N.gif" title="Norte" alt="Norte"></td>
        </tr>

        <tr>
            <th class="borde_rlb_th">(km/h)</th>
            <td colspan="2" class="borde_rb">18&nbsp;</td>
            <td class="borde_b">14&nbsp;</td>
            <td class="borde_rb">14&nbsp;</td>
            <td class="borde_b">14&nbsp;</td>
            <td class="borde_rb">14&nbsp;</td>
            <td class="borde_rb">14&nbsp;</td>
            <td class="borde_rb">11&nbsp;</td>
            <td class="borde_rb">14&nbsp;</td>
            <td class="borde_rb">29&nbsp;</td>
        </tr>

        <tr>
            <th class="borde_rlb_th">Indice UV Máximo</th>
            <td colspan="2" class="borde_rb"><span class="raduv_pred_nivel2" title="índice ultravioleta moderado">&nbsp;&nbsp;4</span></td>
            <td colspan="2" class="borde_rb"><span class="raduv_pred_nivel2" title="índice ultravioleta moderado">&nbsp;&nbsp;4</span></td>
            <td colspan="2" class="borde_rb"><span class="raduv_pred_nivel2" title="índice ultravioleta moderado">&nbsp;&nbsp;5</span></td>
            <td class="borde_rb">&nbsp;</td>
            <td class="borde_rb">&nbsp;</td>
            <td class="borde_rb">&nbsp;</td>
            <td class="borde_rb">&nbsp;</td>
        </tr>
    </tbody>
</table>
"""

########################################################
# Parsers
########################################################

def parse_table_head(thead):
    """Obtiene un array de fechas a partir de la cabecera de la tabla.
    """
    dates_array = []
    row = thead.find('tr', 'cabecera_niv1')
    for cell in row.findAll('th')[1:]:
        dates_array.append(elem2str(cell))
    return dates_array
    
def parse_table_body(tbody):
    """Obtiene un array de celdas con el cuerpo de la tabla.
    """
    table_array = []
    for row in tbody.findAll('tr', recursive=False):
        row_array = []
        row_array.append(elem2str(row.find('th')))

        for cell in row.findAll('td', recursive=False):
            contents = cell.contents[0]
            row_array.append(contents)
            if cell.has_key('colspan'):
                row_array.append(contents)
        
        table_array.append(row_array)
    
    return table_array

########################################################
# Decoders
########################################################

def decode_forecast(dates, data):
    forecast = []
    
    for i in range(7):
        forecast.append({
            'fecha'      : dates[i],
            'estado_img' : (),
            'estado_alt' : (),
            'prob_prec'  : -1,
            'temp_max'   : -1,
            'temp_min'   : -1,
            'viento_img' : (),
            'viento_alt' : (),
            'viento_kmh' : (),
            'indice_uv'  : -1,
        })
    
    for row in data:
        name = row[0].lower()
        day = 0
        
        if name.find(u'cielo') != -1:
            for i in range(1,7,2):
                forecast[day]['estado_img'] = (decode_image_src(row[i]),
                                               decode_image_src(row[i+1]))
                forecast[day]['estado_alt'] = (decode_image_alt(row[i]),
                                               decode_image_alt(row[i+1]))
                day += 1
            for i in range(7,11):
                forecast[day]['estado_img'] = (decode_image_src(row[i]),)
                forecast[day]['estado_alt'] = (decode_image_alt(row[i]),)
                day += 1

        elif name.find(u'precip') != -1:
            for i in (1,3,5,7,8,9,10):
                forecast[day]['prob_prec'] = elem2str(row[i])
                day += 1
    
        elif name.find(u'máxima') != -1:
            for i in (1,3,5,7,8,9,10):
                forecast[day]['temp_max'] = elem2str(row[i])
                day += 1
    
        elif name.find(u'mínima') != -1:
            for i in (1,3,5,7,8,9,10):
                forecast[day]['temp_min'] = elem2str(row[i])
                day += 1
    
        elif name.find(u'viento') != -1:
            for i in range(1,7,2):
                forecast[day]['viento_img'] = (decode_image_src(row[i]),
                                               decode_image_src(row[i+1]))
                forecast[day]['viento_alt'] = (decode_image_alt(row[i]),
                                               decode_image_alt(row[i+1]))
                day += 1
            for i in range(7,11):
                forecast[day]['viento_img'] = (decode_image_src(row[i]),)
                forecast[day]['viento_alt'] = (decode_image_alt(row[i]),)
                day += 1

        elif name.find(u'km/h') != -1:
            for i in range(1,7,2):
                forecast[day]['viento_kmh'] = (elem2str(row[i]),
                                               elem2str(row[i+1]))
                day += 1
            for i in range(7,11):
                forecast[day]['viento_kmh'] = (elem2str(row[i]),)
                day +=1
    
        elif name.find(u'uv') != -1:
            for i in (1,3,5,7,8,9,10):
                forecast[day]['indice_uv'] = elem2str(row[i])
                day += 1
                
    return forecast


########################################################
# xxx
########################################################

def elem2str(elem):
    if isinstance(elem, Tag):
        contents = elem.find(text=True).string
    else:
        contents = str(elem)
    return re.sub('[\s\xc2\xa0]+', ' ', contents).strip()


########################################################
# Decoders
########################################################

def decode_image_src(img):
	if img and img.has_key('src'):
		src = img['src']
		res = re.search("^.*/(.*?)$", src)
		if res:
		    return res.group(1)
	return ""

def decode_image_alt(img):
	if img and img.has_key('alt'):
		return img['alt']
	return ""

def decode_location(title):
    location = elem2str(title)
    res = re.search(".*\. (.*) (.*)", location)
    if res:
        location = res.groups()
        res = re.search("(.*), (.*)", location[0])
        if res:
            groups = res.groups()
            return ("%s %s" % (groups[1], groups[0]), location[1])
        else:
            return location
    else:
        return location


########################################################
# 
########################################################

def connect_to_server(location_code, language, timeout):
    """Devuelve un stream que es necesario cerrarlo después de usarlo
    """
    url = SERVER_URL % (language, location_code)

    data_stream = None
    
    try:
        if timeout > 0:
            oldtimeout = socket.getdefaulttimeout()
            socket.setdefaulttimeout(timeout)

        try:
            data_stream = urllib2.urlopen(url)
        except urllib2.URLError, e:
            raise IOError, e
        except Exception:
            error_message = "Failed contacting server."
            raise RuntimeError, error_message
    finally:
        if timeout > 0:
            socket.setdefaulttimeout(oldtimeout)
            
    return data_stream

    
########################################################
# Public API
########################################################

def local_weather(location_code, language='es', path_src='meteo_icons/', timeout=0):
    """Devuelve un dict con la predicción de la web AEMET
    """
    # Connecto to server
    data_stream = connect_to_server(location_code, language, timeout)

    # Process HTML with BeautifulSoup
    soup = BeautifulSoup(data_stream,
                         convertEntities=BeautifulSoup.HTML_ENTITIES)

    # Close data stream
    data_stream.close()

    location = ("", "")    # Localidad, Provincia
    capital  = ""          # Localidad (altitud)
    date     = ""          # Fecha de elaboración
    forecast = []          # Predicción
    forecast_html = ""

    try:
        # Extract location
        title = soup.find('h2', 'titulo')
        if title is not None:
            location = decode_location(title)
        
        # Extract capital, date and HTML
        notes = soup.find('div', 'notas_tabla')
        if notes is not None:
            spans = notes.findAll('span')
            if len(spans) == 2:
                capital = elem2str(spans[0].nextSibling)
                date    = elem2str(spans[1].nextSibling)
    
            # HTML
            forecast_html = notes.parent
            forecast_html['class'] = 'meteoDetails'
    
            for img in forecast_html.findAll('img'):
                img['src'] = path_src + decode_image_src(img)

        # Extract forecast
        table = soup.find('table', 'tabla_datos')

        thead = table.thead
        dates = parse_table_head(thead)
        
        tbody = table.tbody
        data  = parse_table_body(tbody)
        
        forecast = decode_forecast(dates, data)
    
    except:
        error_message = "Failed parsing HTML."
        raise RuntimeError, error_message
 
    return {
        "location" : location,
        "capital"  : capital,
        "date"     : date,
        "forecast" : forecast,
        "table"    : str(forecast_html),
    }
